src/mlpack/tests/augmented_rnns_tasks_test.cpp

```diff
@@ -0,0 +1,274 @@
+// the required number of times
+// (yes, no ML here, we're unit testing :)
+class HardCodedCopyModel {
+ public:
+  HardCodedCopyModel() : nRepeats(1) {}
+  void Train(
+      arma::field<arma::mat>& predictors,
+      arma::field<arma::mat>& labels) {
+    auto input = predictors.at(0);
+    auto output = labels.at(0);
+    size_t zeroCnt = 0, oneCnt = 0;
+    for (size_t i = 1; i < input.n_rows; i += 2) {
+      size_t& addVar = (input.at(i, 0) == 0) ? zeroCnt : oneCnt;
+      ++addVar;
+    }
+    assert(oneCnt % zeroCnt == 0);
+    nRepeats = oneCnt / zeroCnt;
+  }
+  void Predict(
+      arma::mat& predictors,
+      arma::mat& labels) {
+    size_t seqLen = (predictors.n_rows / 2) / (nRepeats + 1);
+    size_t outputLen = nRepeats * seqLen;
+    assert(2 * (seqLen + outputLen) == predictors.n_rows);
+    labels.zeros(predictors.n_rows / 2, 1);
+    for (size_t i = 0; i < outputLen; ++i) {
+      labels.at(seqLen+i) = predictors.at(2 * (i % seqLen));
+    }
+  }
+  void Predict(
+      arma::field<arma::mat>& predictors,
+      arma::field<arma::mat>& labels) {
+    size_t sz = predictors.n_elem;
+    labels = arma::field<arma::mat>(sz);
+    for (size_t i = 0; i < sz; ++i) {
+      Predict(predictors.at(i), labels.at(i));
+    }
+  }
+
+ private:
+  size_t nRepeats;
+};
+
+class HardCodedSortModel {
+ public:
+  HardCodedSortModel() {}
+  void Train(arma::field<arma::mat>& predictors,
+             arma::field<arma::mat>& labels)
+  {
+    assert(predictors.n_elem == labels.n_elem);
+    bitLen = predictors.at(0).n_rows;
+  }
+  void Predict(
+      arma::mat& predictors,
+      arma::mat& labels)
+  {
+    size_t len = predictors.n_cols;
+    labels.zeros(bitLen, len);
+    vector<pair<int, int>> vals(len);
+    for (size_t j = 0; j < len; ++j) {
+      int val = 0;
+      for (size_t k = 0; k < bitLen; ++k) {
+        val <<= 1;
+        val += predictors.at(k, j);
+      }
+      vals[j] = make_pair(val, j);
+    }
+    sort(vals.begin(), vals.end());
+    for (size_t j = 0; j < len; ++j) {
+      labels.col(j) = predictors.col(vals[j].second);
+    }
+  }
+  void Predict(arma::field<arma::mat>& predictors,
+               arma::field<arma::mat>& labels) {
+    auto sz = predictors.n_elem;
+    labels = arma::field<arma::mat>(sz);
+    for (size_t i = 0; i < sz; ++i) {
+      Predict(predictors.at(i), labels.at(i));
+    }
+  }
+
+ private:
+  size_t bitLen;
+};
+
+class HardCodedAddModel {
+ public:
+  HardCodedAddModel() {}
+  void Train(arma::field<arma::mat>& predictors,
+             arma::field<arma::mat>& labels)
+  {
+    return;
+  }
+  void Predict(arma::mat& predictors,
+               arma::mat& labels)
+  {
+    assert(predictors.n_rows == 3);
+    int num_A = 0, num_B = 0;
+    bool num = false; // true iff we have already seen the separating symbol
+    size_t len = predictors.n_cols;
+    for (size_t i = 0; i < len; ++i) {
+      auto digit = arma::as_scalar(arma::find(1 == predictors.col(i), 1));
```

Comment:
Just as an FYI it is typically good to avoid `auto` when using Armadillo (I think in this particular situation it works ok but not in general). This is because the types returned by Armadillo expressions are actually placeholder types that represent the mathematical operation to be performed, and these are not meant to be instant images. I can clarify more if you like, but I wanted to point it out.